### PR TITLE
top: Add command `top`

### LINKED ISSUE #84: Add command `top`
could i implement ps and top for this, also there already exist projects like bottom

### PR DESCRIPTION
**Pull request by Krysztal112233**
Fix #84

### CODE REVIEW

**Comment by Krysztal112233** [comment]
This pr only implemented output, and the TUI will implemented in other pr as same as `slabtop`

**Comment by Krysztal112233** [comment]
There are more problems with the current pr, but solutions do not yet exist for these problems.

- There should be a word like `rt` in `PR` field, but the getpriority function does not provide a corresponding response.
- The value of the `PR` field looks incorrect, the `top` command output should be 20 mostly, but the current implementation mostly outputs 0 or 19.

**Comment by cakebaker** [comment]
I guess you have seen that `test_flag_user` fails on Windows?

File: tests/by-util/test_top.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+#[test]
+fn test_conflict_arg() {
+    new_ucmd!().arg("-p 0").arg("-U 0").fails().code_is(1);
+}
+
+#[test]
+fn test_flag_user() {
+    let binding = new_ucmd!().arg("-U=root").succeeds();
+    let output = binding.code_is(0).stderr_str();
```
Comment (cakebaker) [inline tests/by-util/test_top.rs:21]: Why do you use `stderr_str()` and not `stdout_str()`?

File: src/uu/top/src/top.rs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+        Self { filter, width }
+    }
+}
+
+#[uucore::main]
+pub fn uumain(args: impl uucore::Args) -> UResult<()> {
+    let matches = uu_app().try_get_matches_from(args)?;
+
+    // Must refresh twice.
```
Comment (cakebaker) [inline src/uu/top/src/top.rs:60]: Why?

File: src/uu/top/src/top.rs [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    fn new(matches: &ArgMatches) -> Self {
+        let filter = if let Some(pidlist) = matches.get_many::<u32>("pid") {
+            Some(Filter::Pid(pidlist.cloned().collect::<Vec<_>>()))
+        } else if let Some(user) = matches.get_one::<String>("filter-any-user") {
+            Some(Filter::User(user.clone()))
+        } else {
+            matches
+                .get_one::<String>("filter-only-euser")
+                .map(|euser| Filter::EUser(euser.clone()))
+        };
```
Comment (cakebaker) [inline src/uu/top/src/top.rs:48]: It looks a bit odd to use two different ways to accomplish the same. Maybe it would make sense to add another `if let`.
```rust
        } else if let Some(user) = matches.get_one::<String>("filter-any-user") {
            Some(Filter::User(user.clone()))
        } else if let Some(euser) = matches.get_one::<String>("filter-only-euser") {
            Some(Filter::EUser(euser.clone()))
        } else {
            None
        };
```
Another idea is to get rid of all the `if`s and use `map` and `or_else`.
```rust
        let filter = matches
            .get_many::<u32>("pid")
            .map(|pidlist| Filter::Pid(pidlist.cloned().collect()))
            .or_else(|| {
                matches
                    .get_one::<String>("filter-any-user")
                    .map(|user| Filter::User(user.clone()))
            })
            .or_else(|| {
                matches
                    .get_one::<String>("filter-only-euser")
                    .map(|euser| Filter::EUser(euser.clone()))
            });
```

File: src/uu/top/src/top.rs [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    let settings = Settings::new(&matches);
+
+    if let Some(Filter::User(user)) = &settings.filter {
+        if Users::new_with_refreshed_list()
+            .iter()
+            .all(|it| it.name() != user)
+        {
+            return Err(USimpleError::new(1, "top: Invalid user"));
+        }
+    }
```
Comment (cakebaker) [inline src/uu/top/src/top.rs:74]: This works only partially. When using `--filter-any-user=0`, an "invalid user" error is shown whereas the original `top` shows no error because it treats numbers as uids.

The handling of `--filter-only-euser` should probably be similar to the handling of `--filter-any-user`.

File: src/uu/top/src/top.rs [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    picker::sysinfo().write().unwrap().refresh_all();
+
+    let settings = Settings::new(&matches);
+
+    if let Some(Filter::User(user)) = &settings.filter {
+        if Users::new_with_refreshed_list()
+            .iter()
+            .all(|it| it.name() != user)
+        {
+            return Err(USimpleError::new(1, "top: Invalid user"));
```
Comment (cakebaker) [inline src/uu/top/src/top.rs:72]: `USimpleError` automatically prefixes the error message with the tool name, so currently the error message shown is "top: top: Invalid user".
```suggestion
            return Err(USimpleError::new(1, "Invalid user"));
```

File: src/uu/top/src/top.rs (hunk 2 above)
Comment (Krysztal112233) [reply src/uu/top/src/top.rs:60]: If we want to fetch CPU usage of process, we have to fresh twice, it's required by sysinfo crate.

[sysinfo](https://docs.rs/sysinfo/0.31.2/sysinfo/struct.System.html#method.refresh_cpu_usage)

File: tests/by-util/test_top.rs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        new_ucmd!()
+            .arg("-U=root")
+            .succeeds()
+            .code_is(0)
+            .stdout_str(),
+    );
+    check(new_ucmd!().arg("-U=0").succeeds().code_is(0).stdout_str());
+
+    new_ucmd!().arg("-U=NOT_EXIST").fails().code_is(1);
+    new_ucmd!().arg("-U=19999").fails().code_is(1);
```
Comment (cakebaker) [inline tests/by-util/test_top.rs:40]: I think this is not correct. When running `-U=19999` with the original `top`, it doesn't fail. It simply shows the header and no processes.

File: src/uu/top/src/top.rs (hunk 2 above)
Comment (cakebaker) [reply src/uu/top/src/top.rs:64]: Can you please add this link to the comment?

File: src/uu/top/src/top.rs [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+            // arg!(-d  --delay                <SECS>          "iterative delay as SECS [.TENTHS]"),
+            // arg!(-E  --"scale-summary-mem"  <SCALE>         "set mem as: k,m,g,t,p,e for SCALE"),
+            // arg!(-e  --"scale-task-mem"     <SCALE>         "set mem with: k,m,g,t,p for SCALE"),
+            // arg!(-H  --"threads-show"                       "show tasks plus all their threads"),
+            // arg!(-i  --"idle-toggle"                        "reverse last remembered 'i' state"),
+            // arg!(-n  --iterations           <NUMBER>        "exit on maximum iterations NUMBER"),
+            arg!(-O  --"list-fields"                        "output all field names, then exit"),
+            // arg!(-o  --"sort-override"      <FIELD>         "force sorting on this named FIELD"),
+            arg!(-p  --pid                  <PIDLIST>       "monitor only the tasks in PIDLIST")
+                .action(ArgAction::Append),
```
Comment (cakebaker) [inline src/uu/top/src/top.rs:266]: I think a value parser is missing here. When running `cargo run -q top -p <PID>` I get the following error:
```
thread 'main' panicked at src/uu/top/src/top.rs:41:14:
Mismatch between definition and access of `pid`. Could not downcast to u32, need to downcast to alloc::string::String
```

File: src/uu/top/src/picker.rs [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    let f = |cmd: &[OsString]| -> String {
+        let binding = cmd
+            .iter()
+            .map(|os_str| os_str.to_string_lossy().into_owned())
+            .collect::<Vec<_>>()
+            .join(" ");
+        let trimmed = binding.trim();
+
+        if trimmed.is_empty() {
+            "[kthreadd]".into()
```
Comment (cakebaker) [inline src/uu/top/src/picker.rs:161]: Why do you use `[kthreadd]`? The original `top` doesn't use `[kthreadd]`:
```
26     root              0    TODO  TODO  TODO  TODO  S  0.00   0.0   8:12.49  [kthreadd] 
57     root              0    TODO  TODO  TODO  TODO  S  0.00   0.0   8:12.49  [kthreadd]

26 root     -51   0       0      0      0 S   0.0   0.0   0:00.00 idle_inject/2                                                                               
57 root     -51   0       0      0      0 S   0.0   0.0   0:01.55 irq/9-acpi
```

File: tests/by-util/test_top.rs (hunk 6 above)
Comment (Krysztal112233) [reply tests/by-util/test_top.rs:40]: Fixed 

```shell
❯ cargo -q run top -U1999
TODO


 PID  USER  PR  NI  VIRT  RES  SHR  S  %CPU  %MEM  TIME+  COMMAND 
```

File: src/uu/top/src/top.rs (hunk 7 above)
Comment (Krysztal112233) [reply src/uu/top/src/top.rs:267]: Fixed 

```shell
❯ cargo run -q top -p 1
TODO


 PID  USER  PR  NI    VIRT  RES   SHR   S  %CPU  %MEM  TIME+    COMMAND 
 1    root  0   TODO  TODO  TODO  TODO  S  0.00  0.0   2:19.31  /sbin/init splash 
```

File: src/uu/top/src/top.rs [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+
+#[derive(Debug)]
+struct Settings {
+    // batch:bool
+    filter: Option<Filter>,
+    width: Option<usize>,
+}
+
+impl Settings {
+    fn new(matches: &ArgMatches) -> UResult<Self> {
```
Comment (cakebaker) [inline src/uu/top/src/top.rs:39]: Why did you change the return type from `Self` to `UResult<Self>`?

File: src/uu/top/src/top.rs [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+            .or_else(|| {
+                matches
+                    .get_one::<String>("filter-only-euser")
+                    .map(|euser| Filter::EUser(euser.clone()))
+            });
+
+        let filter = match filter {
+            Some(Filter::User(data)) => Some(Filter::User(try_into_uid(data)?)),
+            // TODO: Make sure this working
+            Some(Filter::EUser(data)) => Some(Filter::User(try_into_uid(data)?)),
```
Comment (cakebaker) [inline src/uu/top/src/top.rs:79]: I guess it should be `EUser` in both cases:
```suggestion
            Some(Filter::EUser(data)) => Some(Filter::EUser(try_into_uid(data)?)),
```

File: src/uu/top/src/top.rs
Comment (cakebaker) [inline src/uu/top/src/top.rs:1]: I'm not sure whether you are aware of it: the recent changes broke `--filter-any-user`. When running `cargo run top --filter-any-user=root`, it no longer shows any processes.

Comment (Krysztal112233) [reply src/uu/top/src/top.rs:1]: Yes, I'll fix it
